node/cn/handler_test.go

```diff
@@ -58,3 +62,169 @@ func TestNewProtocolManager(t *testing.T) {
 		assert.Equal(t, errIncompatibleConfig, err)
 	}
 }
+
+func TestSampleSize(t *testing.T) {
+	peers := make([]Peer, minNumPeersToSendBlock-1)
+	assert.Equal(t, len(peers), sampleSize(peers))
+
+	peers = make([]Peer, 4)
+	assert.Equal(t, minNumPeersToSendBlock, sampleSize(peers))
+
+	peers = make([]Peer, 16)
+	assert.Equal(t, 4, sampleSize(peers))
+}
+
+func TestSamplePeersToSendBlock(t *testing.T) {
+	pm := &ProtocolManager{}
+	mockCtrl := gomock.NewController(t)
+
+	block := newBlock(123)
+	hash := block.Hash()
+
+	// 1. When the given node is a consensus node.
+	{
+		// len(pnsWithoutBlock) <= blockReceivingPNLimit
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.CONSENSUSNODE
+		pm.peers = mockPeers
+
+		mockCN := NewMockPeer(mockCtrl)
+		mockPN := NewMockPeer(mockCtrl)
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return([]Peer{mockPN}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 2, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
+		assert.Equal(t, mockPN, sampledPeers[1])
+
+		// len(pnsWithoutBlock) > blockReceivingPNLimit
+		mockPeers = NewMockPeerSet(mockCtrl)
+		pm.peers = mockPeers
+
+		mockCN = NewMockPeer(mockCtrl)
+		mockPN = NewMockPeer(mockCtrl)
+
+		var pnWithoutBlock []Peer
+		for i := 0; i <= blockReceivingPNLimit; i++ {
+			pnWithoutBlock = append(pnWithoutBlock, mockPN)
+		}
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
+		mockPeers.EXPECT().PNWithoutBlock(hash).Return(pnWithoutBlock).Times(1)
+
+		sampledPeers = pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1+blockReceivingPNLimit, len(sampledPeers))
+		assert.Equal(t, mockCN, sampledPeers[0])
+	}
+
+	// 2. When the given node is a proxy node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.PROXYNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().PeersWithoutBlockExceptCN(hash).Return([]Peer{mockPeer}).Times(1)
+
+		sampledPeers := pm.samplePeersToSendBlock(block)
+		assert.Equal(t, 1, len(sampledPeers))
+		assert.Equal(t, mockPeer, sampledPeers[0])
+	}
+
+	// 3. When the given node is an end point node.
+	{
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.ENDPOINTNODE
+		pm.peers = mockPeers
+
+		mockPeer := NewMockPeer(mockCtrl)
+		mockPeers.EXPECT().ENWithoutBlock(hash).Return([]Peer{mockPeer}).Times(1)
```

Comment:
In order to confirm that EN transmits only to EN, it is likely to make the same setting for PNWithoutBlock and CNWithoutBlock and check whether it is not included in peer list.